File: python/history_store.py
# history_store.py
"""Istoricul offload-urilor efectuate — port 1:1 al `HistoryStore.swift`
(Mac), care la randul lui e portat din DataMover. Persistat in
%AppData%\\CGConvertor\\ (Windows), intre lansari ale aplicatiei."""

import json
import os
import platform
from datetime import datetime

MAX_ENTRIES = 200


def _app_support_dir():
    if platform.system() == "Windows":
        base = os.environ.get("APPDATA", os.path.expanduser("~"))
    else:
        base = os.path.expanduser("~/Library/Application Support")
    path = os.path.join(base, "CGConvertor")
    os.makedirs(path, exist_ok=True)
    return path


class HistoryStore:
    def __init__(self):
        self.path = os.path.join(_app_support_dir(), "offload_history.json")
        self.entries = self._load()

    def _load(self):
        try:
            with open(self.path, "r", encoding="utf-8") as f:
                return json.load(f)
        except (OSError, json.JSONDecodeError):
            return []

    def _save(self):
        try:
            with open(self.path, "w", encoding="utf-8") as f:
                json.dump(self.entries, f, ensure_ascii=False, indent=2)
        except OSError:
            pass

    def record(self, folder_name, source_path, destination_targets, ok_count, mismatch_count, error_count):
        entry = {
            "date_text": datetime.now().strftime("%d.%m.%Y %H:%M"),
            "folder_name": folder_name,
            "sources_summary": os.path.basename(source_path),
            "dest_summary": ", ".join(os.path.basename(p) for p in destination_targets),
            "ok_count": ok_count, "mismatch_count": mismatch_count, "error_count": error_count,
            "source_paths": [source_path], "destination_target_paths": destination_targets,
        }
        self.entries.append(entry)
        if len(self.entries) > MAX_ENTRIES:
            self.entries = self.entries[-MAX_ENTRIES:]
        self._save()

    def delete(self, index):
        if 0 <= index < len(self.entries):
            del self.entries[index]
            self._save()

    def clear_all(self):
        self.entries = []
        self._save()


_shared = None


def shared():
    global _shared
    if _shared is None:
        _shared = HistoryStore()
    return _shared

File: CGConvertor/HistoryStore.swift
import Foundation
import Combine

/// Istoricul offload-urilor efectuate — port din DataMover
/// (`HistoryStore.swift`): data, folderul creat, sursele/destinațiile
/// folosite, câte fișiere OK/nepotrivire/eroare. Persistat în Application
/// Support, între lansări ale aplicației.
struct HistoryEntry: Codable, Identifiable {
    var id: String { "\(dateText)-\(folderName)" }
    let dateText: String
    let folderName: String
    let sourcesSummary: String
    let destSummary: String
    let okCount: Int
    let mismatchCount: Int
    let errorCount: Int
    var sourcePaths: [String] = []
    var destinationTargetPaths: [String] = []
}

final class HistoryStore: ObservableObject {
    static let shared = HistoryStore()

    private let fileURL: URL
    @Published private(set) var entries: [HistoryEntry] = []

    private init() {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            .appendingPathComponent("CGConvertor", isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        fileURL = base.appendingPathComponent("offload_history.json")
        load()
    }

    private func load() {
        guard let data = try? Data(contentsOf: fileURL),
              let decoded = try? JSONDecoder().decode([HistoryEntry].self, from: data) else { return }
        entries = decoded
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }

    func record(folderName: String, sourcePath: String, destinationTargets: [String],
                okCount: Int, mismatchCount: Int, errorCount: Int) {
        let df = DateFormatter()
        df.dateFormat = "dd.MM.yyyy HH:mm"
        let entry = HistoryEntry(
            dateText: df.string(from: Date()), folderName: folderName,
            sourcesSummary: (sourcePath as NSString).lastPathComponent,
            destSummary: destinationTargets.map { ($0 as NSString).lastPathComponent }.joined(separator: ", "),
            okCount: okCount, mismatchCount: mismatchCount, errorCount: errorCount,
            sourcePaths: [sourcePath], destinationTargetPaths: destinationTargets
        )
        entries.append(entry)
        if entries.count > 200 { entries.removeFirst(entries.count - 200) }
        save()
    }

    func delete(_ entry: HistoryEntry) {
        entries.removeAll { $0.id == entry.id }
        save()
    }

    func clearAll() {
        entries.removeAll()
        save()
    }
}
